File: code-example-tests/java/driver-sync/src/main/java/timeseries/SecondaryIndexes.java
package timeseries;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.CreateCollectionOptions;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Indexes;
import com.mongodb.client.model.TimeSeriesOptions;
import com.mongodb.client.model.geojson.Point;
import com.mongodb.client.model.geojson.Position;
import org.bson.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class SecondaryIndexes {
    private static final String uri = System.getenv("CONNECTION_STRING") != null
            ? System.getenv("CONNECTION_STRING")
            : "Env variable not found. Verify you have a .env file with a valid connection string.";

    private static final String databaseName = "timeseries";
    private static final String collectionName = "sensorData";

    private static MongoCollection<Document> collection;
    private static MongoDatabase database;
    private static MongoClient client;

    private static void loadData() {
        client = MongoClients.create(uri);
        database = client.getDatabase(databaseName);

        // :snippet-start: secondary-create-collection
        CreateCollectionOptions createCollectionOptions = new CreateCollectionOptions()
                .timeSeriesOptions(
                        new TimeSeriesOptions("timestamp")
                                .metaField("metadata"))
                .expireAfter(24, TimeUnit.HOURS);
        // :snippet-end:
        database.createCollection(collectionName, createCollectionOptions);
        collection = database.getCollection(collectionName);

        // :snippet-start: secondary-data
        List<Document> sampleDocuments = Arrays.asList(
                new Document()
                        // :snippet-start: secondary-data-schema
                        .append("metadata", new Document()
                                .append("sensorId", 5578)
                                .append("type", "omni")
                                .append("location", new Document()
                                        .append("type", "Point")
                                        .append("coordinates", Arrays.asList(-77.40711, 39.03335))))
                        // :snippet-end:
                        .append("timestamp", Date.from(Instant.parse("2022-01-15T00:00:00Z")))
                        .append("currentConditions", new Document()
                                .append("windDirection", 127.0)
                                .append("tempF", 71.0)
                                .append("windSpeed", 2.0)
                                .append("cloudCover", null)
                                .append("precip", 0.1)
                                .append("humidity", 94.0)),
                new Document()
                        .append("metadata", new Document()
                                .append("sensorId", 5578)
                                .append("type", "omni")
                                .append("location", new Document()
                                        .append("type", "Point")
                                        .append("coordinates", Arrays.asList(-77.40711, 39.03335))))
                        .append("timestamp", Date.from(Instant.parse("2022-01-15T00:01:00Z")))
                        .append("currentConditions", new Document()
                                .append("windDirection", 128.0)
                                .append("tempF", 69.8)
                                .append("windSpeed", 2.2)
                                .append("cloudCover", null)
                                .append("precip", 0.1)
                                .append("humidity", 94.3)),
                new Document()
                        .append("metadata", new Document()
                                .append("sensorId", 5579)
                                .append("type", "omni")
                                .append("location", new Document()
                                        .append("type", "Point")
                                        .append("coordinates", Arrays.asList(-80.19773, 25.77481))))
                        .append("timestamp", Date.from(Instant.parse("2022-01-15T00:01:00Z")))
                        .append("currentConditions", new Document()
                                .append("windDirection", 115.0)
                                .append("tempF", 88.0)
                                .append("windSpeed", 1.0)
                                .append("cloudCover", null)
                                .append("precip", 0.0)
                                .append("humidity", 99.0))
        );
        // :snippet-end:
        collection.insertMany(sampleDocuments);
    }

    public static class SecondaryIndexResult {
        public List<Document> result;
        public Document explainResult;

        public SecondaryIndexResult(List<Document> result, Document explainResult) {
            this.result = result;
            this.explainResult = explainResult;
        }
    }

    public static class CompoundIndexResult {
        public List<Document> result;
        public Document explainResult;
        public List<Document> hintResult;

        public CompoundIndexResult(List<Document> result, Document explainResult, List<Document> hintResult) {
            this.result = result;
            this.explainResult = explainResult;
            this.hintResult = hintResult;
        }
    }

    public static SecondaryIndexResult createAndUseSecondaryIndex() {
        loadData();
        // :snippet-start: simple-in-example
        var filter = Filters.in("metadata.type", "temperature", "pressure");
        // :snippet-end:

        // :snippet-start: create-secondary-index
        collection.createIndex(Indexes.ascending("timestamp"));
        // :snippet-end:

        // :snippet-start: sort-with-secondary-index
        var matchFilter = Filters.gte("timestamp",
                Date.from(Instant.parse("2022-01-15T00:00:00Z")));

        var pipeline = Arrays.asList(
                new Document("$match", new Document("timestamp",
                        new Document("$gte", Date.from(Instant.parse("2022-01-15T00:00:00Z"))))),
                new Document("$sort", new Document("timestamp", 1))
        );

        List<Document> result = collection.aggregate(pipeline).into(new ArrayList<>());
        // :snippet-end:

        // :snippet-start: sort-with-secondary-index-explain
        var explainCommand = new Document("explain", new Document()
                .append("aggregate", collectionName)
                .append("pipeline", pipeline)
                .append("cursor", new Document()))
                .append("verbosity", "executionStats");

        Document explainResult = database.runCommand(explainCommand);
        // :snippet-end:

        return new SecondaryIndexResult(result, explainResult);
    }

    public static CompoundIndexResult createAndUseCompoundIndexes() {
        // :snippet-start: last-point-indexes
        // Indexes on ``timeField`` descending are more performant because they
        // enable ``DISTINCT_SCAN`` optimizations.
        List<String> indexes = Arrays.asList(
                collection.createIndex(Indexes.compoundIndex(
                        Indexes.ascending("metadata.sensorId"),
                        Indexes.ascending("timestamp"))),
                collection.createIndex(Indexes.compoundIndex(
                        Indexes.ascending("metadata.sensorId"),
                        Indexes.descending("timestamp"))),
                collection.createIndex(Indexes.compoundIndex(
                        Indexes.descending("metadata.sensorId"),
                        Indexes.ascending("timestamp"))),
                collection.createIndex(Indexes.compoundIndex(
                        Indexes.descending("metadata.sensorId"),
                        Indexes.descending("timestamp")))
        );
        // :snippet-end:

        // :snippet-start: last-point-index-meta-up-time-down
        collection.createIndex(Indexes.compoundIndex(
                Indexes.ascending("metadata.type"),
                Indexes.descending("timestamp")));
        // :snippet-end:

        // :snippet-start: sort-and-group
        var pipeline = Arrays.asList(
                new Document("$sort", new Document()
                        .append("metadata.sensorId", 1)
                        .append("timestamp", -1)),
                new Document("$group", new Document()
                        .append("_id", "$metadata.sensorId")
                        .append("ts", new Document("$first", "$timestamp"))
                        .append("temperatureF", new Document("$first", "$currentConditions.tempF")))
        );

        List<Document> result = collection.aggregate(pipeline).into(new ArrayList<>());
        // :snippet-end:

        // :snippet-start: sort-and-group-explain
        var explainCommand = new Document("explain", new Document()
                .append("aggregate", collectionName)
                .append("pipeline", pipeline)
                .append("cursor", new Document()))
                .append("verbosity", "executionStats");

        Document explainResult = database.runCommand(explainCommand);
        // :snippet-end:

        // :snippet-start: hint
        var hintPipeline = Arrays.asList(
                new Document("$sort", new Document()
                        .append("metadata.sensorId", 1)
                        .append("timestamp", -1)),
                new Document("$group", new Document()
                        .append("_id", "$metadata.sensorId")
                        .append("ts", new Document("$first", "$timestamp"))
                        .append("temperatureF", new Document("$first", "$currentConditions.tempF")))
        );

        List<Document> hintResult = collection.aggregate(hintPipeline)
                .hint(new Document()
                        .append("metadata.sensorId", 1)
                        .append("timestamp", -1))
                .into(new ArrayList<>());
        // :snippet-end:

        return new CompoundIndexResult(result, explainResult, hintResult);
    }

    public static List<Document> createAndUseGeospatialIndex() {
        // :snippet-start: create-geospatial-index-temp
        collection.createIndex(Indexes.geo2dsphere("currentConditions.tempF"));
        // :snippet-end:

        // :snippet-start: create-geospatial-index-location
        collection.createIndex(Indexes.geo2dsphere("metadata.location"));
        // :snippet-end:

        var filter = Filters.nearSphere(
                "metadata.location",
                new Point(new Position(-77.03653, 38.897676)),
                5000000.0,
                null);

        List<Document> result = collection.find(filter).into(new ArrayList<>());

        return result;
    }

    public static void cleanup() {
        collection.drop();
        client.close();
    }
}

File: code-example-tests/javascript/driver/examples/time-series/secondary-indexes/secondary-indexes.js
import { Collection, MongoClient } from 'mongodb';

const uri =
  process.env.CONNECTION_STRING ||
  'Env variable not found. Verify you have a .env file with a valid connection string.';

const databaseName = 'timeseries';
const collectionName = 'sensorData';

let client;
let database;
let collection;

async function loadData() {
  // Close existing client if any to avoid connection leaks
  if (client) {
    await client.close();
  }
  client = new MongoClient(uri);
  await client.connect();
  database = client.db(databaseName);

  // :snippet-start: secondary-create-collection
  const createCollectionOptions = {
    timeseries: {
      timeField: 'timestamp',
      metaField: 'metadata',
    },
    expireAfterSeconds: 86400,
  };
  // :snippet-end:

  await database.createCollection(collectionName, createCollectionOptions);
  collection = database.collection(collectionName);

  // :snippet-start: secondary-data
  const sampleDocuments = [
    {
      // :snippet-start: secondary-data-schema
      metadata: {
        sensorId: 5578,
        type: 'omni',
        location: {
          type: 'Point',
          coordinates: [-77.40711, 39.03335],
        },
      },
      // :snippet-end:
      timestamp: new Date('2022-01-15T00:00:00.000Z'),
      currentConditions: {
        windDirection: 127.0,
        tempF: 71.0,
        windSpeed: 2.0,
        cloudCover: null,
        precip: 0.1,
        humidity: 94.0,
      },
    },
    {
      metadata: {
        sensorId: 5578,
        type: 'omni',
        location: {
          type: 'Point',
          coordinates: [-77.40711, 39.03335],
        },
      },
      timestamp: new Date('2022-01-15T00:01:00.000Z'),
      currentConditions: {
        windDirection: 128.0,
        tempF: 69.8,
        windSpeed: 2.2,
        cloudCover: null,
        precip: 0.1,
        humidity: 94.3,
      },
    },
    {
      metadata: {
        sensorId: 5579,
        type: 'omni',
        location: {
          type: 'Point',
          coordinates: [-80.19773, 25.77481],
        },
      },
      timestamp: new Date('2022-01-15T00:01:00.000Z'),
      currentConditions: {
        windDirection: 115.0,
        tempF: 88.0,
        windSpeed: 1.0,
        cloudCover: null,
        precip: 0.0,
        humidity: 99.0,
      },
    },
  ];
  // :snippet-end:

  await collection.insertMany(sampleDocuments);
}

export async function createAndUseSecondaryIndex() {
  await loadData();
  // :snippet-start: simple-in-example
  const filter = { 'metadata.type': { $in: ['temperature', 'pressure'] } };
  // :snippet-end:

  // :snippet-start: create-secondary-index
  await collection.createIndex({ timestamp: 1 });
  // :snippet-end:

  // :snippet-start: sort-with-secondary-index
  const matchStage = {
    $match: { timestamp: { $gte: new Date('2022-01-15T00:00:00.000Z') } },
  };
  const sortStage = { $sort: { timestamp: 1 } };

  const pipeline = [matchStage, sortStage];

  const result = await collection.aggregate(pipeline).toArray();
  // :snippet-end:

  // :snippet-start: sort-with-secondary-index-explain
  const explainResult = await database.command({
    explain: {
      aggregate: collectionName,
      pipeline: pipeline,
      cursor: {},
    },
    verbosity: 'executionStats',
  });
  // :snippet-end:

  return { result, explainResult };
}

export async function createAndUseCompoundIndexes() {
  // :snippet-start: last-point-indexes
  // Indexes on ``timeField`` descending are more performant because they
  // enable ``DISTINCT_SCAN`` optimizations.
  const indexes = [
    { key: { 'metadata.sensorId': 1, timestamp: 1 } },
    { key: { 'metadata.sensorId': 1, timestamp: -1 } },
    { key: { 'metadata.sensorId': -1, timestamp: 1 } },
    { key: { 'metadata.sensorId': -1, timestamp: -1 } },
  ];
  // :snippet-end:

  for (const index of indexes) {
    await collection.createIndex(index.key);
  }

  // :snippet-start: last-point-index-meta-up-time-down
  await collection.createIndex({ 'metadata.type': 1, timestamp: -1 });
  // :snippet-end:

  // :snippet-start: sort-and-group
  const pipeline = [
    { $sort: { 'metadata.sensorId': 1, timestamp: -1 } },
    {
      $group: {
        _id: '$metadata.sensorId',
        ts: { $first: '$timestamp' },
        temperatureF: { $first: '$currentConditions.tempF' },
      },
    },
  ];

  const result = await collection.aggregate(pipeline).toArray();
  // :snippet-end:

  // :snippet-start: sort-and-group-explain
  const explainResult = await database.command({
    explain: {
      aggregate: collectionName,
      pipeline: pipeline,
      cursor: {},
    },
    verbosity: 'executionStats',
  });
  // :snippet-end:

  // :snippet-start: hint
  const hintPipeline = [
    { $sort: { 'metadata.sensorId': 1, timestamp: -1 } },
    {
      $group: {
        _id: '$metadata.sensorId',
        ts: { $first: '$timestamp' },
        temperatureF: { $first: '$currentConditions.tempF' },
      },
    },
  ];

  const hintResult = await collection
    .aggregate(hintPipeline, {
      hint: { 'metadata.sensorId': 1, timestamp: -1 },
    })
    .toArray();
  // :snippet-end:

  return { result, explainResult, hintResult };
}

export async function createAndUseGeospatialIndex() {
  // :snippet-start: create-geospatial-index-location
  await collection.createIndex({ 'metadata.location': '2dsphere' });
  // :snippet-end:

  const pipeline = [
    {
      $geoNear: {
        near: { type: 'Point', coordinates: [-77.03653, 38.897676] },
        distanceField: 'distance',
        maxDistance: 5000000, // 5 million meters (5,000 km)
        spherical: true,
        key: 'metadata.location',
      },
    },
  ];
  const result = await collection.aggregate(pipeline).toArray();

  return result;
}

export async function cleanup() {
  await database?.dropCollection(collectionName);
  await client?.close();
}
